Encoder/DefaultEncoder.php

```diff
@@ -0,0 +1,163 @@
+<?php
+
+namespace Lexik\Bundle\JWTAuthenticationBundle\Encoder;
+
+use InvalidArgumentException;
+use Lexik\Bundle\JWTAuthenticationBundle\Services\KeyLoader\KeyLoaderInterface;
+use Namshi\JOSE\JWS;
+
+/**
+ * Default class for encode/decode JWT, based on the Namshi/JOSE library.
+ * Supports OpenSSL and PHPSecLib encryption engines.
+ *
+ * @since 2.0
+ * @author Robin Chalas <robin.chalas@gmail.com>
+ */
+class DefaultEncoder extends JWTEncoder
+{
+    /**
+     * @var KeyLoaderInterface
+     */
+    protected $keyLoader;
+
+    /**
+     * @var string
+     */
+    protected $encryptionEngine;
+
+    /**
+     * @var string
+     */
+    protected $encryptionAlgorithm;
+
+    /**
+     * Constructor.
+     *
+     * @param KeyLoaderInterface $keyLoader
+     * @param string             $encryptionEngine
+     * @param string             $encryptionAlgorithm
+     *
+     * @throws InvalidArgumentException If the given encryptionEngine is not supported.
+     * @throws InvalidArgumentException If the given algorithm is not supported.
+     */
+    public function __construct(KeyLoaderInterface $keyLoader, $encryptionEngine, $encryptionAlgorithm)
+    {
+        if (!in_array($encryptionEngine, ['OpenSSL', 'SecLib'])) {
+            throw new InvalidArgumentException(
+                sprintf('The encryption engine "%s" is not supported', $encryptionEngine)
+            );
+        }
+
+        if (!$this->isAlgorithmSupportedForEngine($encryptionEngine, $encryptionAlgorithm)) {
+            throw new InvalidArgumentException(
+                sprintf('The algorithm "%s" is not supported for %s', $encryptionAlgorithm, $encryptionEngine)
+            );
+        }
+
+        $this->keyLoader           = $keyLoader;
+        $this->encryptionEngine    = $encryptionEngine;
+        $this->encryptionAlgorithm = $encryptionAlgorithm;
+    }
+
+    /**
+     * {@inheritdoc}
+     *
+     * @throws \RuntimeException If the JWS cannot be signed
+     */
+    public function encode(array $data)
+    {
+        $jws = new JWS(['alg' => $this->encryptionAlgorithm], $this->encryptionEngine);
```

Comment:
@weaverryan This is the most difficult decision of this changes, I needed some suggestions to do take the good one, so thank's for your help. 

Having two services is effectively not necessary. If I add an `encoder_engine` config option (The `encoder_algorithm` one is used to choose between the algorithms available for each engine, e.g. RS256, ES256, ...), we could safely have only one `lexik_jwt_authentication.encoder.default`. 
I'll go in this way (add this additional `encoder_engine` option, default to "OpenSSL"), custom encoders can use it or not, such as done for the `encoder_algorithm`.
Then, I just have to replace the two services by one and inject the value of the `encoder_engine` option as the second argument (`$encryptionEngine`) of the service.

About use the JWS as a provider and inject it in the encoder service, I'm not sure that's necessary because deal with the `encoder_engine` should be sufficient, as pretty much all can be done with the JWS class in the new DefaultEncoder (it just takes the encryptionEngine as constructor argument, it's already in place and managed with the two different services because of the missing `encoder_engine` option).
I propose you to add the  `encoder_engine` option and be back, then if you think that could make sense to do it, you could talk me more about this way sort as I can try to implement it.